### ISSUE TITLE
Non-contiguous sub.big.matrix?

### ISSUE THREAD

**Issue by cdeterman** [kaneplusplus/bigmemory#20]
Currently we have `sub.big.matrix` but it explicitly states that non-contiguous subsets are not possible.  It would be great if we could find a way to get this implemented before the next release.

**Comment by kaneplusplus** [comment]
Agreed. The ```sub.big.matrix``` was implemented with row and column offsets in the ```BigMatrix``` class. We could probably do a lot more by implementing a new accessor class. However, we need to make sure that the data structure doing the mapping from matrix coordinate to memory location is not so complex that it uses a lot of memory and we need some way to make it clear that it won't work with bigalgebra.

**Comment by cdeterman** [comment]
@kaneplusplus why is it that it won't work with bigalgebra?  That seems like a pretty large constraint if you want to do anything with the data other than some minimal summary statistics.

**Comment by kaneplusplus** [comment]
bigalgebra assumes BLAS/LAPACK and both take contiguous column-major c-arrays as input. If your ```sub.big.matrix``` is simply a subset of the columns then there is no problem since the memory is contiguous. Otherwise you end up memory that is not contiguous.

A quick fix would be to have bigalgebra check to see if the object is a ```sub.big.matrix``` create an anonymous ```big.matrix``` and pass that off to the linear algebra routines.

**Comment by privefl** [comment]
Maybe the simpliest thing to do is to provide a deepcopy-like function more than a sub.big.matrix-like function.

I personally use something like:
`X2 <- big.matrix(length(ind.row), length(ind.col), type = "char", shared = FALSE)
  deepcopyPart(X@address, X2@address, ind.row, ind.col)`

where 
`void deepcopyPart(SEXP pBigMat,
                  SEXP pBigMat2,
                  const IntegerVector& rowInd,
                  const IntegerVector& colInd) {

  XPtr<BigMatrix> xpMat(pBigMat);
  MatrixAccessor\<char> macc(*xpMat);
  XPtr<BigMatrix> xpMat2(pBigMat2);
  MatrixAccessor\<char> macc2(*xpMat2);

  int n = rowInd.size();
  int m = colInd.size();

  int indi, indj;

  for (int j = 0; j < m; j++) {
    indj = colInd[j] - 1;
    for (int i = 0; i < n; i++) {
      indi = rowInd[i] - 1;
      macc2[j][i] = macc[indj][indi];
    }
  }

  return;
}`

Of course, this needs to be more generic (types, cache, etc) but I use it all the time.

By the way, thanks for providing bigmemory, I'm a huge fan of this package !

Florian

**Comment by kaneplusplus** [comment]
Thanks very much!

Please note that deepcopy does let you specify columns and rows to copy from a big.matrix object.

**Comment by cdeterman** [comment]
@privefl Thanks for jumping in, contributions are always welcome.  However as noted by Mike this is available via `deepcopy`.  The reason I created this issue was it would be ideal to have a method to reference non-contiguous elements within a `big.matrix` object WITHOUT copying the elements to another object.

**Comment by privefl** [comment]
I'm very sorry, I read the documentation too quickly, I thought `rows` and `cols` stood for sizes, not indices. 
I will try this option right now.
Thank you both.
